src/main/java/com/books/service/BookService.java

```diff
@@ -25,63 +27,74 @@ public BookService(BookRepository repository) {
 
     /**
      * Persists a book object.
+     *
      * @param book book to be saved.
      */
-    public void saveBook(Book book){
+    public void saveBook(Book book) {
         repository.save(book);
     }
 
     /**
-     * Retrieves all books.
-     * @return all books.
+     * Retrieves all books in the current page with requested page size.
+     *
+     * @param pageNum Page number of request - zero based.
+     * @param entriesPerPage Number of entries per page.
+     * @return All books in the current page with requested page size.
      */
-    public List<Book> getAllBooks(){
-        return (List<Book>) repository.findAll();
+    public List<Book> getAllBooks(int pageNum, int entriesPerPage) {
+        Pageable pageable = PageRequest.of(pageNum, entriesPerPage);
+        return repository.findAll(pageable).toList();
     }
 
     /**
      * Retrieves book by id.
+     *
      * @param id the id of the book to be retrieved.
      * @return the book with the given id.
      */
-    public Optional<Book> getBookById(int id){//negative error check and null check
+    public Optional<Book> getBookById(int id) {//negative error check and null check
         return repository.findById(id);
     }
 
     /**
      * Retrieves all books, which contain a specific substring in it's title.
+     *
      * @param titleContaining the title substring, on which books will be retrieved.
      * @return books with the matching title substring.
      */
-    public List<Book> getBookByTitleContaining(String titleContaining){
-        return repository.findByTitleContaining(titleContaining);
+    public List<Book> getBookByTitleContaining(String titleContaining,int pageNum, int entriesPerPage) {
+        Pageable pageable = PageRequest.of(pageNum, entriesPerPage);
+        return repository.findByTitleContaining(titleContaining,pageable);
     }
 
     /**
      * Deletes a book from the repository.
+     *
      * @param id the id of the book to be deleted.
      */
-    public void deleteBook(int id){
+    public void deleteBook(int id) {
         repository.deleteById(id);
     }
 
     /**
      * Adds rating to a book
+     *
      * @param id the id of the book, whose rating will be increased
      */
-    public void addRating(int id){
-        modifyRating(id, + 1);
+    public void addRating(int id) {
+        modifyRating(id, +1);
     }
 
     /**
      * Removes rating from a book
+     *
      * @param id the id of the book, whose rating will be decreased
      */
-    public void removeRating(int id){
-        modifyRating(id, - 1);
+    public void removeRating(int id) {
```

Comment:
you are not removing the rating for this book, you are decreasing the rating. Use better name.

Reply:
Done